// n8n-nodes-cronlytic/dist/credentials/CronlyticApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CronlyticApi = void 0;
class CronlyticApi {
    constructor() {
        this.name = 'cronlyticApi';
        this.displayName = 'Cronlytic API';
        this.documentationUrl = 'https://www.cronlytic.com/api-documentation';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                default: '',
                required: true,
                typeOptions: {
                    password: true,
                },
                description: 'Your Cronlytic API key',
            },
            {
                displayName: 'User ID',
                name: 'userId',
                type: 'string',
                default: '',
                required: true,
                description: 'Your Cronlytic User ID',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-API-Key': '={{$credentials.apiKey}}',
                    'X-User-ID': '={{$credentials.userId}}',
                    'Content-Type': 'application/json',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.cronlytic.com/prog',
                url: '/ping',
                method: 'GET',
            },
        };
    }
}
exports.CronlyticApi = CronlyticApi;
